### PR TITLE
Add abseil string helpers

### PR DESCRIPTION
**Pull request by asl**
Add abseil string manipulator helpers for
  - cstring
  - IR::Node
  - IR::ID

This improves the code itself as we no longer need to do explicit .toString() / .string_view() calls. Also, this prints directly to sink, providing some performance improvements.

### CODE REVIEW

**Comment by asl** [comment]
This required few breaking changes though:
  - As there is custom formatter, things cannot be formatted via `%s` depending on implicit conversion to `char*`
  - There is an implicit conversion to `std::string_view` now producing some breakage with Protobuf code as now we are having ambiguous implicit conversions. Unfortunately, we cannot (yet) make conversion to `char*` explicit, as lots of code relies on this, and this will be a topic for another PR

**Comment by fruffy** [comment]
Nice, maybe we should wait for #4964 to be merged to avoid compatibility issues?

**Comment by asl** [comment]
> Nice, maybe we should wait for #4964 to be merged to avoid compatibility issues?

I not quite sure. Even more, given the size of the backend, doing project-wide refactoring and changes could be prohibitely expensive as soon as it will be merged in.

**Comment by asl** [comment]
Tofino failures are due to mentioned protobuf ambiguities. They will likely will be easier to fix after the subsequent PR that removes implicit conversion to `char*`.

**Comment by fruffy** [comment]
> Tofino failures are due to mentioned protobuf ambiguities. They will likely will be easier to fix after the subsequent PR that removes implicit conversion to `char*`.

What are the planned changes for the next PR? Maybe you can stack it on this one so we can see?

File: backends/dpdk/control-plane/bfruntime_arch_handler.h
Comment (fruffy) [inline backends/dpdk/control-plane/bfruntime_arch_handler.h:150]: Why is this required here?

File: lib/cstring.h
Comment (fruffy) [inline lib/cstring.h:194]: Would add a comment here why this shouldn't be explicit.

File: backends/dpdk/control-plane/bfruntime_arch_handler.h
Comment (asl) [reply backends/dpdk/control-plane/bfruntime_arch_handler.h:150]: As I mentioned, Protobuf strings have two constructors: from `const char*` and `string_view`. In order to use `cstring` directly we need to make this conversion explicit.

File: lib/cstring.h
Comment (asl) [reply lib/cstring.h:194]: This was already discussed in https://github.com/p4lang/p4c/issues/4688. But essentially: we are going to make interface similar to `std::string` essentially allowing `string_view` into corresponding cstring-backed storage. This would fix lots of `cstring` misuse.

File: backends/dpdk/control-plane/bfruntime_arch_handler.h
Comment (fruffy) [reply backends/dpdk/control-plane/bfruntime_arch_handler.h:150]: And I assume the reason is that we also have an implicit conversion to `const char*`? Is the follow-up PR working on that? Maybe it should be first.

Comment (asl) [reply backends/dpdk/control-plane/bfruntime_arch_handler.h:150]: Right. And this is misused in many places, e.g. passing `cstring's` to C functions like `strlen`, improper checks for nulls, etc. :)

Comment (fruffy) [reply backends/dpdk/control-plane/bfruntime_arch_handler.h]: The changes paired with the subsequent Pr look much better to me. Definitely an improvement.

Comment (asl) [reply backends/dpdk/control-plane/bfruntime_arch_handler.h]: I might missed few cases here and there. However, I believe that we do not need lambdas in the example like in  https://github.com/p4lang/p4c/pull/4951#discussion_r1807919236 anymore.

**Comment by asl** [comment]
@fruffy @vlstill @ChrisDodd Will you please take a look when you will have a chance?

**Comment by fruffy** [comment]
> @fruffy @vlstill @ChrisDodd Will you please take a look when you will have a chance?

Changes generally look good to me, I am waiting for comments by others since this is a breaking change. The Tofino breakage looks easy enough to fix. I will also push some more Tofino fixes later which could help this PR.

**Comment by ChrisDodd** [comment]
We really need multiple formatters for things like IDs and Nodes:

- one for use in error messages, where we output things in a form that the P4 programmer can understand
- one for use in logging messages, where we output things that the compiler developer needs to know in order to debug the compiler.

In the past, we've used `toString()` for the former and `operator <<` for the latter.  That means that the `error` and `warning` functions (and their helpers) will use `toString()` when converting things for `%` formats in the message, while `LOG` macros all use `<<`.

Where does abseil formatting fit in this?  Is it for user messages or logging?  Whatever the choice, it needs to be clearly documented.

**Comment by asl** [comment]
> Where does abseil formatting fit in this? Is it for user messages or logging? Whatever the choice, it needs to be clearly documented.

This simplifies `toString` calls. `operator<<` are untouched (though many of those delegate to some `toString` calls, yes). The main intention here is to simplify explicit `string_view` calls when used in different APIs that are being converted to `string_view`'s these days.

The converters are to simplify `absl::StrCat` / `absl::StrFormat` / `absl::StrSubstitute` / `absl::StrAppend` / `absl::StrJoin` calls, e.g.
```diff
-        std::string tcActionParam = absl::StrCat("\n\tparam ", paramName.string_view(), " type ");
+        std::string tcActionParam = absl::StrCat("\n\tparam ", paramName, " type ");
```

Or, instead of
```c++
name = absl::StrCat(absl::StrJoin(stack, ".", [](std::string *out, cstring s) {
                                 absl::StrAppend(out, s.string_view());
                             }), name.string_view())
```
we end with
```c++
name = absl::StrCat(absl::StrJoin(stack, "."), name);
```

There are all different ways to "convert" `cstring` or `ID` for printing in the codebase: `c_str()` calls, `toString()` calls, `string_view()` calls. I tried to clean the things as much as I can :)

I will add comment describing the intended usage.

**Comment by ChrisDodd** [comment]
> There are all different ways to "convert" `cstring` or `ID` for printing in the codebase: `c_str()` calls, `toString()` calls, `string_view()` calls. I tried to clean the things as much as I can :)
> 
For an `ID` at least, `toString()` and `operator<<` do very different things -- `toString()` give you the ID as it appears in the P4 source code, while `operrator<<` gives you the uniquified/mangled name (so different versions of the same ID in different scopes will look different).

**Comment by asl** [comment]
> For an `ID` at least, `toString()` and `operator<<` do very different things -- `toString()` give you the ID as it appears in the P4 source code, while `operrator<<` gives you the uniquified/mangled name (so different versions of the same ID in different scopes will look different).

That's correct. But neither user-side error reporting, nor debug printing are affected / changed / or even touched. These are just helpers to simplify the usage of abseil string manipulation functions. So, I might just put wrong title of the PR.

**Review by ChrisDodd — APPROVED** [review]

**Pull request merged**
